docs/web/howto/web-in-60/websockets.rst

```diff
@@ -0,0 +1,91 @@
+:LastChangedDate: $LastChangedDate$
+:LastChangedRevision: $LastChangedRevision$
+:LastChangedBy: $LastChangedBy$
+
+WebSockets
+==========
+
+Twisted Web provides support for the WebSocket protocol for clients (such as web browsers) to connect to resources and receive bi-directional communication with the server.
+
+For the purposes of our example here, you will need to have some familiarity with the concepts covered in :doc:`serving static content <static-content>` and :doc:`rpy scripts <rpy-scripts>`, because we will be using those to construct our resource hierarchy.
+
+.. note::
+
+   In order to use the websocket support in Twisted, you will need the ``websocket`` optional dependency, so :doc:`install Twisted </installation>` with either ``pip install twisted[websocket]`` or install one that includes it, such as ``twisted[all_non_platform]`` or ``twisted[all]`` .
+
+WebSocket Server
+----------------
+
+Let's build a demo of a simple websocket server communicating with a web browser.
+To begin with, we will need a folder with 3 files in it.
+First, let's do the Twisted part.  We need a :py:class:`twisted.web.websocket.WebSocketResource` to be served at a known URL, so let's put one into a ``.rpy`` file called ``webskt.rpy`` :
+
+:download:`webskt.rpy <../../examples/websocket/webskt.rpy>`
+
+.. literalinclude:: ../../examples/websocket/webskt.rpy
+
+Note that ``WebSocketEchoFactory`` complies with the :py:class:`twisted.web.websocket.WebSocketServerFactory` :py:class:`protocol <typing.Protocol>`, returning a ``WebSocketDemo`` that complies with :py:class:`twisted.web.websocket.WebSocketProtocol` .
```

Comment:
See my inline comments from the example Python code.

I hope we can have the source code with enough documentation, so that we will no longer need this narrative text.

Reply:
Making every narrative example contain a full but un-indexed copy of the reference documentation seems like a really bad stylistic choice to me. Do you find it hard to navigate to the API docs via these links? I wonder if there is a missing pydoctor feature here

Reply:
I think that the links to pydoctor pages are ok.

But I think that the websocket part should be outside of the twisted-web in 60 seconds pages.

The page is quite deep: Twisted main docs -> Twisted Web -> Developers Guide -> Twisted Web in 60 Seconds -> webSocket


I was expecting to have someting like this: Twisted main docs -> Twisted Web -> WebSocket

------

Something for another ticket ... I think that all developers guides should be move up level up  and have "Examples" as part of "Other developer guides"

Reply:
Yes, the docs really need a big restructuring :( I intentionally did not touch that here because it's a big information architecture project and I wanted to just fit in with the existing (bad) structure.  But this is something we should look at soon.